contrib/gdiffmk/gdiffmk.1.man
.TH gdiffmk @MAN1EXT@ "@MDATE@" "groff @VERSION@"
.SH Name
gdiffmk \- mark differences between
.IR groff / nroff / troff
files
.
.
.\" ====================================================================
.\" Legal Terms
.\" ====================================================================
.\"
.\" Copyright (C) 2004-2020 Free Software Foundation, Inc.
.\"
.\" This file is part of gdiffmk, which is part of groff, the GNU roff
.\" type-setting system.
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see
.\" <http://www.gnu.org/licenses/>.
.
.
.\" Save and disable compatibility mode (for, e.g., Solaris 10/11).
.do nr *groff_gdiffmk_1_man_C \n[.cp]
.cp 0
.
.\" Define fallback for groff 1.23's MR macro if the system lacks it.
.nr do-fallback 0
.if !\n(.f           .nr do-fallback 1 \" mandoc
.if  \n(.g .if !d MR .nr do-fallback 1 \" older groff
.if !\n(.g           .nr do-fallback 1 \" non-groff *roff
.if \n[do-fallback]  \{\
.  de MR
.    ie \\n(.$=1 \
.      I \%\\$1
.    el \
.      IR \%\\$1 (\\$2)\\$3
.  .
.\}
.rr do-fallback
.
.
.\" ====================================================================
.SH Synopsis
.\" ====================================================================
.
.SY gdiffmk
.RB [ \-a\~\c
.IR add-mark ]
.RB [ \-c\~\c
.IR change-mark ]
.RB [ \-d\~\c
.IR delete-mark ]
.RB [ \-x\~\c
.IR diff-command ]
.RB [ \-D
.RB [ \-B ]
.RB [ \-M
.IR "mark1 mark2" ]]
.RB [ \-\- ]
.I file1
.I file2
.RI [ output ]
.YS
.
.
.SY gdiffmk
.B \-\-help
.YS
.
.
.SY gdiffmk
.B \%\-\-version
.YS
.
.
.\" ====================================================================
.SH Description
.\" ====================================================================
.
.I gdiffmk
compares two
.MR roff @MAN7EXT@
documents,
.I file1
and
.IR file2 ,
and creates a
.I roff
document consisting of
.I file2
with added margin character
.RB ( .mc )
requests indicating output lines that differ from
.I file1.
.
If the
.I file1
or
.I file2
argument is
.RB \[lq] \- \[rq],
.I gdiffmk
reads the standard input stream for that input.
.
If the
.I output
operand is present,
.I gdiffmk
writes output to a file of that name.
.
If it is
.RB \[lq] \- \[rq]
or absent,
.I gdiffmk
writes output to the standard output stream.
.
.RB \[lq] \- \[rq]
cannot be both an input and output operand.
.
.
.\" ====================================================================
.SH Options
.\" ====================================================================
.
.B \-\-help
displays a usage message
and
.B \%\-\-version
shows version information;
both exit afterward.
.
.
.TP
.BI \-a\~ add-mark
Use
.I add-mark
for source lines not in
.I file1
but present in
.IR file2 .
.
Default:
.RB \[lq] + \[rq].
.
.
.TP
.B \-B
By default,
the deleted texts marked by the
.B \-D
option end with an added
.I roff
break request,
.BR .br ,
to ensure that the deletions are marked properly.
.
This is the only way to guarantee that deletions and small
changes get flagged.
.
This option directs the program not to insert these breaks;
it makes no sense to use it without
.BR \-D .
.
.
.TP
.BI \-c\~ change-mark
Use
.I change-mark
for changed source lines.
.
Default:
.RB \[lq] | \[rq].
.
.
.TP
.BI \-d\~ delete-mark
Use the
.I delete-mark
for deleted source lines.
.
Default:
.RB \[lq] * \[rq].
.
.TP
.B \-D
Show the deleted portions from changed and deleted text.
.
.
.TP
.BI \-M\~ "mark1 mark2"
Change the delimiting marks for the
.B \-D
option.
.
It makes no sense to use this option without
.BR \-D .
.
Default delimiting marks:
.RB \[lq] [[ "\[rq] .\|.\|.\& \[lq]" ]] \[rq].
.
.
.TP
.BI \-x\~ diff-command
Use the
.I diff-command
command to perform the comparison of
.I file1
and
.IR file2 .
.
In particular,
.I diff-command
should accept the GNU
.MR diff 1
.B \-D
option.
.
Default:
.BR diff .
.
.
.TP
.B \-\-
Treat all subsequent arguments as file names,
even if they begin with
.RB \[lq] \- \[rq].
.
.
.\" ====================================================================
.SH "Exit status"
.\" ====================================================================
.
.I gdiffmk
exits with
.RB status\~ 0
if the input files are the same;
.B 1
if they differ;
.B 2
upon a usage error;
.B 3
if the system's
.MR diff 1
or
.MR sh 1
commands do not support features
.I gdiffmk
requires;
and
.B 4
if the
.I output
argument is a duplicate of
.I file1
or
.IR file2 .
.
.\" ====================================================================
.SH Bugs
.\" ====================================================================
.
The output is not necessarily compatible with all macro packages
and all preprocessors.
.
A workaround that often overcomes preprocessor problems is to run
.I gdiffmk
on the output of all the preprocessors instead of the input source.
.
.
.LP
.I gdiffmk
relies on the
.B \-D
option of GNU
.I diff
to make a merged \[lq]#ifdef\[rq] output format.
.
Busybox
.I diff
is known to not support it.
.
Also see the
.BI \-x\~ diff-command
option.
.
.
.\" ====================================================================
.SH Authors
.\" ====================================================================
.
.I gdiffmk
was written by
.MT MBianchi@\:Foveal\:.com
Mike Bianchi
.ME ,
now retired.
.
It is maintained by the
.I groff
developers.
.
.
.\" ====================================================================
.SH "See also"
.\" ====================================================================
.
.MR groff @MAN1EXT@ ,
.MR nroff @MAN1EXT@ ,
.MR gtroff @MAN1EXT@ ,
.MR roff @MAN7EXT@ ,
.MR diff @MAN1EXT@
.
.
.\" Restore compatibility mode (for, e.g., Solaris 10/11).
.cp \n[*groff_gdiffmk_1_man_C]
.do rr *groff_gdiffmk_1_man_C
.
.
.\" Local Variables:
.\" fill-column: 72
.\" mode: nroff
.\" End:
.\" vim: set filetype=groff textwidth=72:


contrib/gdiffmk/gdiffmk.sh
#!@BASH_PROG@
# Copyright (C) 2004-2020 Free Software Foundation, Inc.
# Written by Mike Bianchi <MBianchi@Foveal.com <mailto:MBianchi@Foveal.com>>
# Thanks to Peter Bray for debugging.

# This file is part of the gdiffmk utility, which is part of groff.

# groff is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# groff is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
# License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# This file is part of GNU gdiffmk.


CMD=`basename $0`

Diagnose () {
	echo >&2 "${CMD}: $@"
}

Usage () {
	status=0
	if test $# -gt 0
	then
		Diagnose "usage error: $@"
		exec 2>&1
		echo
		status=2
	fi
	cat >&2 <<EOF
usage: ${CMD} [ OPTIONS ] FILE1 FILE2 [ OUTPUT ]
Place difference marks into the new version of a groff/nroff/troff document.
FILE1 and FILE2 are compared, using 'diff', and FILE2 is output with
groff '.mc' requests added to indicate how it is different from FILE1.

  FILE1   Previous version of the groff file.  '-' means standard input.
  FILE2   Current version of the groff file.   '-' means standard input.
          Either FILE1 or FILE2 can be standard input, but not both.
  OUTPUT  Copy of FILE2 with '.mc' commands added.
          '-' means standard output (the default).
          If the shell's 'test' does not support option -ef, OUTPUT
          can only be the standard output.

OPTIONS:
  -a ADDMARK     Mark for added groff source lines.    Default: '+'.
  -c CHANGEMARK  Mark for changed groff source lines.  Default: '|'.
  -d DELETEMARK  Mark for deleted groff source lines.  Default: '*'.

  -D             Show the deleted portions from changed and deleted text.
                  Default delimiting marks:  '[[' .... ']]'.
  -B             By default, the deleted texts marked by the '-D' option end
                  with an added troff '.br' command.  This option prevents
                  the added '.br'.
  -M MARK1 MARK2 Change the delimiting marks for the '-D' option.

  -x DIFFCMD     Use a different diff(1) command;
                  one that accepts the '-Dname' option, such as GNU diff.
  -s SEDCMD      Use a different sed(1) command;
                  such as GNU sed.
  --version      Print version information on the standard output and exit.
  --help         Print this message on the standard error.
EOF
	exit $status
}


Exit () {
	exitcode=$1
	shift
	for arg
	do
		echo >&2 "${CMD}: $1"
		shift
	done
	exit ${exitcode}
}

#	Usage:  FileRead  exit_code  filename
#
#	Check for existence and readability of given file name.
#	If not found or not readable, print message and exit with EXIT_CODE.
FileRead () {
	case "$2" in
	-)
		return
		;;
	esac

	if test ! -f "$2"
	then
		Exit $1 "File '$2' not found."
	fi
	if test ! -r "$2"
	then
		Exit $1 "File '$2' not readable."
	fi
}


#	Usage:  FileCreate  exit_code  filename
#
#	Create the given filename if it doesn't exist.
#	If unable to create or write, print message and exit with EXIT_CODE.
FileCreate () {
	case "$2" in
	-)
		return
		;;
	esac

	touch "$2" 2>/dev/null
	if test $? -ne 0
	then
		if test ! -f "$2"
		then
			Exit $1 "File '$2' not created; " \
			  "Cannot write directory '`dirname "$2"`'."
		fi
		Exit $1 "File '$2' not writeable."
	fi
}

WouldClobber () {
	case "$2" in
	-)
		return
		;;
	esac

	# BASH_PROG is set to /bin/sh if bash was not found
	if test "$HAVE_TEST_EF_OPTION" = "no" -a "$BASH_PROG" = "/bin/sh"
	then
		Exit 3 \
		"Your shell does support test -ef, [OUTPUT] can only be the" \
		"standard  output."
	else
		if test "$1" -ef "$3"
		then
			Exit 4 \
			"The $2 and OUTPUT arguments both point to the same file," \
			"'$1', and it would be overwritten."
		fi
	fi
}

ADDMARK='+'
CHANGEMARK='|'
DELETEMARK='*'
MARK1='[['
MARK2=']]'

RequiresArgument () {
	#	Process flags that take either concatenated or
	#	separated values.
	case "$1" in
	-??*)
		expr "$1" : '-.\(.*\)'
		return 1
		;;
	esac

	if test $# -lt 2
	then
		Exit 2 "Option '$1' requires a value."
	fi

	echo "$2"
	return 0
}

HAVE_TEST_EF_OPTION=@HAVE_TEST_EF_OPTION@
BASH_PROG=@BASH_PROG@
BADOPTION=
DIFFCMD=@DIFF_PROG@
SEDCMD=sed
D_option=
br=.br
while [ $# -gt 0 ]
do
	OPTION="$1"
	case "${OPTION}" in
	-a*)
		ADDMARK=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		;;
	-c*)
		CHANGEMARK=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		;;
	-d*)
		DELETEMARK=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		;;
	-D )
		D_option=D_option
		;;
	-M* )
		MARK1=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		if [ $# -lt 2 ]
		then
			Usage "Option '-M' is missing the MARK2 value."
		fi
		MARK2="$2"
		shift
		;;
	-B )
		br=.
		;;
	-s* )
		SEDCMD=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		;;
	-x* )
		DIFFCMD=`RequiresArgument "${OPTION}" "$2"`		&&
			shift
		;;
	--version)
		echo "${CMD} (groff) version @VERSION@"
		exit 0
		;;
	--help)
		Usage
		;;
	--)
		#	What follows  --  are file arguments
		shift
		break
		;;
	-)
		break
		;;
	-*)
		BADOPTION="invalid option '$1'"
		;;
	*)
		break
		;;
	esac
	shift
done

if ! ${DIFFCMD} -Dx /dev/null /dev/null >/dev/null 2>&1
then
	Exit 3 "The '${DIFFCMD}' program does not accept"	\
"the required '-Dname' option.
Use GNU diff instead.  See the '-x DIFFCMD' option.  You can also
install GNU diff as 'gdiff' on your system."
fi

if test -n "${BADOPTION}"
then
	Usage "${BADOPTION}"
fi

if test $# -lt 2  -o  $# -gt 3
then
	Usage "Incorrect number of arguments."
fi

if test "1$1" = "1-"  -a  "2$2" = "2-"
then
	Usage "Both FILE1 and FILE2 are '-'."
fi

FILE1="$1"
FILE2="$2"

FileRead 1 "${FILE1}"
FileRead 2 "${FILE2}"

if test $# = 3
then
	case "$3" in
	-)
		#	output goes to standard output
		;;
	*)
		#	output goes to a file
		WouldClobber "${FILE1}" FILE1 "$3"
		WouldClobber "${FILE2}" FILE2 "$3"

		FileCreate 3 "$3"
		exec >$3
		;;
	esac
fi

#	To make a very unlikely LABEL even more unlikely ...
LABEL=__diffmk_$$__

SED_SCRIPT='
		/^#ifdef '"${LABEL}"'/,/^#endif \/\* '"${LABEL}"'/ {
		  /^#ifdef '"${LABEL}"'/          s/.*/.mc '"${ADDMARK}"'/
		  /^#endif \/\* '"${LABEL}"'/     s/.*/.mc/
		  p
		  d
		}
		/^#ifndef '"${LABEL}"'/,/^#endif \/\* [!not ]*'"${LABEL}"'/ {
		  /^#else \/\* '"${LABEL}"'/,/^#endif \/\* '"${LABEL}"'/ {
		    /^#else \/\* '"${LABEL}"'/    s/.*/.mc '"${CHANGEMARK}"'/
		    /^#endif \/\* '"${LABEL}"'/   s/.*/.mc/
		    p
		    d
		  }
		  /^#endif \/\* [!not ]*'"${LABEL}"'/ {
		   s/.*/.mc '"${DELETEMARK}"'/p
		   a\
.mc
		  }
		  d
		}
		p
	'

if [ ${D_option} ]
then
	SED_SCRIPT='
		/^#ifdef '"${LABEL}"'/,/^#endif \/\* '"${LABEL}"'/ {
		  /^#ifdef '"${LABEL}"'/          s/.*/.mc '"${ADDMARK}"'/
		  /^#endif \/\* '"${LABEL}"'/     s/.*/.mc/
		  p
		  d
		}
		/^#ifndef '"${LABEL}"'/,/^#endif \/\* [!not ]*'"${LABEL}"'/ {
		  /^#ifndef '"${LABEL}"'/ {
		   i\
'"${MARK1}"'
		   d
		  }
		  /^#else \/\* '"${LABEL}"'/ !{
		   /^#endif \/\* [!not ]*'"${LABEL}"'/ !{
		    p
		    d
		   }
		  }
		  /^#else \/\* '"${LABEL}"'/,/^#endif \/\* '"${LABEL}"'/ {
		    /^#else \/\* '"${LABEL}"'/ {
		     i\
'"${MARK2}"'\
'"${br}"'
		     s/.*/.mc '"${CHANGEMARK}"'/
		     a\
.mc '"${CHANGEMARK}"'
		     d
		    }
		    /^#endif \/\* '"${LABEL}"'/   s/.*/.mc/
		    p
		    d
		  }
		  /^#endif \/\* [!not ]*'"${LABEL}"'/ {
		   i\
'"${MARK2}"'\
'"${br}"'
		   s/.*/.mc '"${DELETEMARK}"'/p
		   a\
.mc
		  }
		  d
		}
		p
	'
fi

${DIFFCMD} -D"${LABEL}" -- "${FILE1}" "${FILE2}"  |
	${SEDCMD} -n "${SED_SCRIPT}"

# EOF
